rust/src/bittorrent_dht/logger.rs

```diff
@@ -0,0 +1,89 @@
+/* Copyright (C) 2020 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+use super::bittorrent_dht::BitTorrentDHTTransaction;
+use crate::jsonbuilder::{JsonBuilder, JsonError};
+
+fn log_bittorrent_dht(
+    tx: &BitTorrentDHTTransaction, js: &mut JsonBuilder,
+) -> Result<(), JsonError> {
+    js.set_string("transaction_id", &tx.transaction_id)?;
+    if let Some(client_version) = &tx.client_version {
+        js.set_string("client_version", client_version)?;
+    }
+    if let Some(request_type) = &tx.request_type {
+        js.set_string("request_type", &request_type)?;
+    }
+    if let Some(error) = &tx.error {
+        js.open_object("error")?;
+        js.set_uint("num", u64::from(error.num))?;
+        js.set_string("msg", &error.msg)?;
+        js.close()?;
+    };
+    if let Some(request) = &tx.request {
+        js.open_object("request")?;
+        js.set_string("id", &request.id)?;
+        if let Some(target) = &request.target {
+            js.set_string("target", target)?;
+        }
+        if let Some(info_hash) = &request.info_hash {
+            js.set_string("info_hash", info_hash)?;
+        }
+        if let Some(token) = &request.token {
+            js.set_string("token", token)?;
+        }
+        /* TODO should implied_port & port just be combined into single port
```

Comment:
I guess I would keep the logging of the raw network message, and let post processing compute the result port out of `implied_port`, `port` and the UDP flow port

Reply:
Cool, I was also thinking of just allowing post-processing to compute the result port. Will remove comment. :)